apps/jupyter/extensions/notebook-runner/src/packages.ts

```diff
@@ -0,0 +1,208 @@
+import { JupyterFrontEnd } from "@jupyterlab/application";
+import { ISessionContext } from "@jupyterlab/apputils";
+import { INotebookTracker } from "@jupyterlab/notebook";
+import { IKernelConnection } from "@jupyterlab/services/lib/kernel/kernel";
+import { Contents, ContentsManager, KernelMessage } from "@jupyterlab/services";
+import { addKernelListeners, writeJsonToVirtualFilesystem } from "./utils";
+import { EmbeddedPythonCallbacks } from "./iframe-api";
+
+const autoInstallPackages =
+  (contentsManager: ContentsManager, notebookPath: string) =>
+  async (kernel: IKernelConnection): Promise<void> => {
+    kernel.iopubMessage.connect((_, msg) => {
+      // ignore status messages
+      if (KernelMessage.isStatusMsg(msg)) {
+        return;
+      } else if (KernelMessage.isStreamMsg(msg)) {
+        console.log(`[${kernel.id}]`, msg.content.text);
+        return;
+      } else if (KernelMessage.isExecuteInputMsg(msg)) {
+        console.debug(
+          `[${kernel.id}] Executing code (${msg.content.execution_count}):\n`,
+          msg.content.code,
+        );
+        return;
+      } else if (KernelMessage.isExecuteResultMsg(msg)) {
+        console.debug(
+          `[${kernel.id}] Finished executing code (${msg.content.execution_count}):`,
+          msg.content.data,
+        );
+      }
+
+      console.debug(`[${kernel.id}] IOPub message:`, msg);
```

Comment:
I guess there's no need for the multiple `return;` above, if we do the following:
```suggestion
      else {
        console.debug(`[${kernel.id}] IOPub message:`, msg);
      }
```